File: 2029-stone-game-ix/solution.py
class Solution:
    def stoneGameIX(self, stones: list[int]) -> bool:
        # Count remainders 0, 1, and 2
        cnt0, cnt1, cnt2 = 0, 0, 0
        for s in stones:
            rem = s % 3
            if rem == 0:
                cnt0 += 1
            elif rem == 1:
                cnt1 += 1
            else:
                cnt2 += 1
        
        # Scenario 1: Even number of stones with remainder 0
        # These stones don't change the eventual turn order for 1s and 2s.
        # Alice wins if both Type 1 and Type 2 stones are available.
        if cnt0 % 2 == 0:
            return cnt1 > 0 and cnt2 > 0
        
        # Scenario 2: Odd number of stones with remainder 0
        # Bob can use the 0-stone to flip the turn parity.
        # Alice needs a lead of at least 3 in either Type 1 or Type 2 
        # stones to force a win.
        return abs(cnt1 - cnt2) > 2

File: 2029-stone-game-ix/solution.cpp
class Solution {
public:
    bool stoneGameIX(vector<int>& stones) {
        // Count remainders when divided by 3
        int cnt[3] = {0, 0, 0};
        for (int s : stones) {
            cnt[s % 3]++;
        }

        // If count of 0-remainder stones is even, it doesn't change the turn parity 
        // in the long run. Alice wins if she can start a sequence and Bob is 
        // forced to lose. This requires both Type 1 and Type 2 stones to exist.
        if (cnt[0] % 2 == 0) {
            return cnt[1] > 0 && cnt[2] > 0;
        }

        // If count of 0-remainder stones is odd, Bob can flip the turn.
        // Alice wins only if she has a surplus of one type (more than 2)
        // to offset the turn flip.
        return abs(cnt[1] - cnt[2]) > 2;
    }
};